In [204]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [205]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data = pd.read_csv(Path("Resources/lending_data.csv"))
# Review the DataFrame
lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [206]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = lending_data["loan_status"]

# Separate the X variable, the features
x = lending_data.drop(columns="loan_status")

In [207]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [208]:
# Review the X variable DataFrame
x.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [209]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(x, y, random_state=1, stratify=y)

In [210]:
# Display the shapes of the resulting datasets
print(f"Training data shape: {X_train.shape}, Training labels shape: {y_train.shape}")
print(f"Testing data shape: {X_test.shape}, Testing labels shape: {y_test.shape}")

Training data shape: (58152, 7), Training labels shape: (58152,)
Testing data shape: (19384, 7), Testing labels shape: (19384,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [211]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
LR_model = logistic_regression_model.fit(X_train, y_train)

# Display a message confirming the model has been trained
print("Logistic Regression model has been trained.")

Logistic Regression model has been trained.


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [212]:
# Make a prediction using the testing data
testing_predictions = LR_model.predict(X_test)
print(len(testing_predictions))

19384


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [213]:
# Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, testing_predictions)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[18673    86]
 [   32   593]]


In [214]:
# Print the classification report for the model
class_report = classification_report(y_test, testing_predictions)
print("\nClassification Report:")
print(class_report)


Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.95      0.91       625

    accuracy                           0.99     19384
   macro avg       0.94      0.97      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
The Logistic Regression model demonstrates strong predictive performance for both 0 (healthy loan) and 1 (high-risk loan) labels:
1. Prediction of 0 (Healthy Loan):

* Precision: 1.00 – The model correctly identified all instances of healthy loans with no false positives.    
* Recall: 1.00 – The model successfully identified all actual healthy loans.  
* F1-score: 1.00 – The model achieved perfect balance between precision and recall for healthy loans. 

2. Prediction of 1 (High-Risk Loan):

* Precision: 0.87 – While there are some false positives, the model accurately predicted high-risk loans in 87% of cases.     
* Recall: 0.95 – The model correctly identified 95% of the actual high-risk loans.    
* F1-score: 0.91 – The model achieved a strong balance between precision and recall for high-risk loans, though slightly lower than for healthy loans.    

# Overall Performance:

* The model achieves an impressive overall accuracy of 99%, indicating that it correctly predicts the status of loans in 99% of cases.    
* The macro and weighted averages further confirm the model’s effectiveness, with high scores across precision, recall, and F1 measures.  
    
In summary, the Logistic Regression model performs exceptionally well in predicting healthy loans and is highly effective in identifying high-risk loans, demonstrating robust performance across both classes.

---